### PR TITLE
Update publish script pre-requisite logic & validation phase

### PR DESCRIPTION
**Pull request by sfsholden**
### What does this PR do?
Follow-up items for the publishing script based on last week's release.
1. Add pre-validation step for checking environment variables.
1. Add install steps for required tools.
1. Add additional check for tool installations (Jason was having issues after switching node versions)

### What issues does this PR fix or reference?
@W-7250936@

### CODE REVIEW

File: scripts/validation-utils.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    console.log(`\nVerifying lerna is installed for node version ${NODE_VERSION}.`);
+    if (!shell.which('lerna') || !shell.which('lerna').includes(NODE_VERSION)) {
+      console.log('Lerna not found - Installing lerna version ' + LERNA_VERSION);
+      shell.exec('npm install -g lerna@' + LERNA_VERSION);
+    }
+  },
+
+  checkVSCEInstall: () => {
+    console.log(`\nVerifying vsce is installed for node version ${NODE_VERSION}.`);
+    if (!shell.which('vsce') || !shell.which('vsce').includes(NODE_VERSION)) {
```
Comment (sfsholden) [inline scripts/validation-utils.js]: The issue Jason had run into last week around the vsce installation was due to switching node versions. He had to reinstall vsce for node version 12.4.0.

File: scripts/validation-utils.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env node
```
Comment (sfsholden) [inline scripts/validation-utils.js]: Refactored a lot of reused validations into a util class.

File: scripts/publish-circleci.js [hunk 3]
```diff
@@ -15,83 +26,25 @@ shell.set('+v');
...
-}
-
-if (!shell.which('aws')) {
-  console.log('aws cli is not installed or could not be found');
+if (process.argv.indexOf('-v') > -1) {
+  console.log('Running prevalidation of environment variables.');
+  console.log(`SALESFORCEDX_VSCODE_VERSION: ${process.env['SALESFORCEDX_VSCODE_VERSION']}`);
+  console.log(`CIRCLECI_TOKEN: ${process.env['CIRCLECI_TOKEN']}`);
+  console.log(`CIRCLECI_BUILD: ${process.env['CIRCLECI_BUILD']}`);
   process.exit(-1);
```
Comment (lcampos) [inline scripts/publish-circleci.js]: Should the exit status here be -1 ?

File: scripts/publish-circleci.js [hunk 4]
```diff
@@ -15,83 +26,25 @@ shell.set('+v');
...
-
-if (currentBranch !== releaseBranchName) {
-  console.log(
-    `You must execute this script in a release branch, you are currently running the script on branch ${currentBranch}`
-  );
-  process.exit(-1);
-}
+checkEnvironmentVariables();
+checkBaseBranch(`release/v${process.env['SALESFORCEDX_VSCODE_VERSION']}`);
+checkNodeVerion();
```
Comment (lcampos) [inline scripts/publish-circleci.js]: Nitpick, misspelled version

File: scripts/validation-utils.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    if (!shell.which('vsce') || !shell.which('vsce').includes(NODE_VERSION)) {
+      console.log('VSCE not found - Installing latest version.')
+      shell.exec('npm install -g vsce');
+    }
+  },
+
+  checkAWSCliInstall: () => {
+    console.log(`\nVerifying AWS CLI is installed for node version ${NODE_VERSION}.`);
+    if (!shell.which('aws')) {
+      shell.exec('pip3 install awscli --upgrade --user');
```
Comment (lcampos) [inline scripts/validation-utils.js]: Is pip3 available by default in linux/mac os versions we've been using ?

File: scripts/publish-circleci.js (hunk 3 above)
Comment (sfsholden) [reply scripts/publish-circleci.js]: No, probably not. I will look for one that is more appropriate.

File: scripts/publish-circleci.js (hunk 4 above)
Comment (sfsholden) [reply scripts/publish-circleci.js]: Thank you

File: scripts/validation-utils.js (hunk 1 above)
Comment (sfsholden) [reply scripts/validation-utils.js]: To expand - this checks that the installation of the current vsce matches the version of node that we require.

File: scripts/validation-utils.js (hunk 5 above)
Comment (sfsholden) [reply scripts/validation-utils.js]: It was for me on both machines. I will add the installation steps though in case it's not for someone going forward.

File: scripts/validation-utils.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      if (!shell.which('aws')) {
+        console.log('AWS CLI is not installed or could not be found.');
+        console.log('This could be a path issue. Example of resolution:');
+        console.log('Install happens in: /Users/<user_name>/Library/Python/3.7/lib/...');
+        console.log(`Add 'export PATH=~/Library/Python/3.7/bin/:$PATH' to your ~/.bash_profile`);
```
Comment (sfsholden) [inline scripts/validation-utils.js]: This is a bit wordy, but I wanted to include it to save time for anyone going forward. I don't think it'd be wise to automate this piece since we'd be modifying someone's .bash_profile.

I think in the future this will be replaced with a gitaction.

Comment (lcampos) [reply scripts/validation-utils.js]: Yeah, maybe we should remove this piece that installs and configures the aws cli for now and just document the process.
